lib/Test/NoLeaks.pm

```diff
@@ -0,0 +1,337 @@
+      track_memory  => ...,
+      track_fds     => ...,
+      passes        => ...,
+      warmup_passes => ...,
+    ), "non-leaked code description";
+
+=head1 DESCRIPTION
+
+It is hard to track memory leaks. There are a lot of perl modules (e.g.
+L<Test::LeakTrace>), that try to B<detect> and B<point> leaks. Unfortunately,
+they do not always work, and they are rather limited because they are not
+able to detect leaks in XS-code or external libraries.
+
+Instead of examining perl internals, this module offers a bit naive empirical
+approach: let the suspicious code to be launched in infinite loop
+some time and watch (via tools like C<top>)if the memory consumption by
+perl process increses over time. If it does, while it is expected to
+be constant (stabilized), then, surely, there are leaks.
+
+This approach is able only to B<detect> and not able to B<point> them. The
+module C<Test::NoLeaks> implements the general idea of the approach, which
+might be enough in many cases.
+
+=head1 INTERFACE
+
+=head3 C<< test_noleaks >>
+
+=head3 C<< noleaks >>
+
+The mandatory hash has the following members
+
+=over 2
+
+=item * C<code>
+
+Suspicious for leaks subroutine, that will be executed multiple times.
+
+=item * C<track_memory>
+
+=item * C<track_fds>
+
+Track memory or file descriptor leaks. At leas one of them should be
+specified.
+
+In B<Unices>, every socket is file descriptor too, so, C<track_fds>
+will be able to track unclosed sockets, i.e. network connections.
+
+=item * C<passes>
+
+How many times C<code> should be executed. If memory leak is too small,
+number of passes should be large enough to trigger additional pages
+allocation for perl process, and the leak will be detected.
+
+Page size is 4kb on linux, so, if C<code> leaks 4 bytes on every
+pass, then C<1024> passes should be specified.
+
+In general, the more passes are specified, the more chance to
+detect possible leaks.
+
+Default value is C<100>. Minimal value is C<2>.
+
+=item * C<warmup_passes>
+
+How many times the C<code> should be executed before module starts
+tracking resources consumption on executing the C<code> C<passes>
+times.
+
+If you have caches, memoizes etc., then C<warmup_passes> is your
+friend.
+
+Default value is C<0>.
+
+=item * C<tolerate_hits>
+
+How many passes, which considered leaked, should be ingnored, i.e.
+maximal number of possible false leak reports.
+
+Even if your code has no leaks, it might cause perl interpreter
+allocate additional memory pages, e.g. due to memory fragmentation.
+Those allocations are legal, and should not be treated as leaks.
+
+Default value is C<0>.
+
+=back
+
+=cut
+
+my $PAGE_SIZE;
+
+BEGIN {
+    no strict "subs";
+    *_platform_mem_size = _linux_mem_size;
+    *_platform_fds = _linux_fds;
+    $PAGE_SIZE = sysconf(_SC_PAGESIZE)
+      or die("page size cannot be determined, Test::NoLeaks cannot be used");
+}
+
+sub _linux_mem_size {
+    my $class = shift;
+
+    open(my $fh, '<', '/proc/self/statm')
```

Comment:
if you keep this file open and seek to the beginning instead of reopening it, perl is less likely to fragment memory.